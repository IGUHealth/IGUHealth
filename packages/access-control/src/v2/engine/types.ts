import { FHIRClientAsync } from "@iguhealth/client/lib/interface";
import {
  AllInteractions,
  FHIRRequest,
  FHIRResponse,
} from "@iguhealth/client/lib/types";
import {
  ClientApplication,
  Membership,
  OperationDefinition,
} from "@iguhealth/fhir-types/r4/types";
import { FHIR_VERSION } from "@iguhealth/fhir-types/versions";
import { AccessTokenPayload } from "@iguhealth/jwt";

export type Result<CTX, Role, Result> = {
  context: PolicyContext<CTX, Role>;
  result: Result;
};

export interface PolicyContext<CTX, Role> {
  clientCTX: CTX;
  client: FHIRClientAsync<CTX>;
  environment: {
    request: FHIRRequest<FHIR_VERSION, AllInteractions>;
    user: {
      payload: AccessTokenPayload<Role>;
      resource: Membership | ClientApplication | OperationDefinition;
    };
  };
  attributes: {
    [key: string]: FHIRResponse<FHIR_VERSION, AllInteractions>;
  };
}
