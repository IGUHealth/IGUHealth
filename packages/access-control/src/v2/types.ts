import { FHIRClientAsync } from "@iguhealth/client/lib/interface";
import { FHIRRequest, FHIRResponse } from "@iguhealth/client/lib/types";
import {
  ClientApplication,
  Membership,
  OperationDefinition,
} from "@iguhealth/fhir-types/r4/types";
import { AccessTokenPayload } from "@iguhealth/jwt";

export type Result<CTX, Role, Result> = {
  context: PolicyContext<CTX, Role>;
  result: Result;
};

export interface PolicyContext<CTX, Role> {
  clientCTX: CTX;
  client: FHIRClientAsync<CTX>;
  environment: {
    request: FHIRRequest;
    user: {
      payload: AccessTokenPayload<Role>;
      resource: Membership | ClientApplication | OperationDefinition;
    };
  };
  attributes: {
    [key: string]: FHIRResponse;
  };
}
