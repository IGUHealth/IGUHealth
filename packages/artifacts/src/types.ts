export interface PackageJSON {
  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
}

export type IndexedFiles = {
  filename?: string;
  resourceType?: string;
  id?: string;
  url?: string;
  version?: string;
  kind?: string;
  type?: string;
  supplements?: string;
  content?: string;
};

export type IndexFile = {
  "index-version"?: string;
  files?: Array<IndexedFiles>;
};
