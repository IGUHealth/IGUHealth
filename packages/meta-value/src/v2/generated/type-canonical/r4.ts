// Do not this code is generated 
export default {
  "Resource": "http://hl7.org/fhir/StructureDefinition/Resource",
  "Account": "http://hl7.org/fhir/StructureDefinition/Account",
  "ActivityDefinition": "http://hl7.org/fhir/StructureDefinition/ActivityDefinition",
  "AdverseEvent": "http://hl7.org/fhir/StructureDefinition/AdverseEvent",
  "AllergyIntolerance": "http://hl7.org/fhir/StructureDefinition/AllergyIntolerance",
  "Appointment": "http://hl7.org/fhir/StructureDefinition/Appointment",
  "AppointmentResponse": "http://hl7.org/fhir/StructureDefinition/AppointmentResponse",
  "AuditEvent": "http://hl7.org/fhir/StructureDefinition/AuditEvent",
  "Basic": "http://hl7.org/fhir/StructureDefinition/Basic",
  "Binary": "http://hl7.org/fhir/StructureDefinition/Binary",
  "BiologicallyDerivedProduct": "http://hl7.org/fhir/StructureDefinition/BiologicallyDerivedProduct",
  "BodyStructure": "http://hl7.org/fhir/StructureDefinition/BodyStructure",
  "Bundle": "http://hl7.org/fhir/StructureDefinition/Bundle",
  "CapabilityStatement": "http://hl7.org/fhir/StructureDefinition/CapabilityStatement",
  "CarePlan": "http://hl7.org/fhir/StructureDefinition/CarePlan",
  "CareTeam": "http://hl7.org/fhir/StructureDefinition/CareTeam",
  "CatalogEntry": "http://hl7.org/fhir/StructureDefinition/CatalogEntry",
  "ChargeItem": "http://hl7.org/fhir/StructureDefinition/ChargeItem",
  "ChargeItemDefinition": "http://hl7.org/fhir/StructureDefinition/ChargeItemDefinition",
  "Claim": "http://hl7.org/fhir/StructureDefinition/Claim",
  "ClaimResponse": "http://hl7.org/fhir/StructureDefinition/ClaimResponse",
  "ClinicalImpression": "http://hl7.org/fhir/StructureDefinition/ClinicalImpression",
  "CodeSystem": "http://hl7.org/fhir/StructureDefinition/CodeSystem",
  "Communication": "http://hl7.org/fhir/StructureDefinition/Communication",
  "CommunicationRequest": "http://hl7.org/fhir/StructureDefinition/CommunicationRequest",
  "CompartmentDefinition": "http://hl7.org/fhir/StructureDefinition/CompartmentDefinition",
  "Composition": "http://hl7.org/fhir/StructureDefinition/Composition",
  "ConceptMap": "http://hl7.org/fhir/StructureDefinition/ConceptMap",
  "Condition": "http://hl7.org/fhir/StructureDefinition/Condition",
  "Consent": "http://hl7.org/fhir/StructureDefinition/Consent",
  "Contract": "http://hl7.org/fhir/StructureDefinition/Contract",
  "Coverage": "http://hl7.org/fhir/StructureDefinition/Coverage",
  "CoverageEligibilityRequest": "http://hl7.org/fhir/StructureDefinition/CoverageEligibilityRequest",
  "CoverageEligibilityResponse": "http://hl7.org/fhir/StructureDefinition/CoverageEligibilityResponse",
  "DetectedIssue": "http://hl7.org/fhir/StructureDefinition/DetectedIssue",
  "Device": "http://hl7.org/fhir/StructureDefinition/Device",
  "DeviceDefinition": "http://hl7.org/fhir/StructureDefinition/DeviceDefinition",
  "DeviceMetric": "http://hl7.org/fhir/StructureDefinition/DeviceMetric",
  "DeviceRequest": "http://hl7.org/fhir/StructureDefinition/DeviceRequest",
  "DeviceUseStatement": "http://hl7.org/fhir/StructureDefinition/DeviceUseStatement",
  "DiagnosticReport": "http://hl7.org/fhir/StructureDefinition/DiagnosticReport",
  "DocumentManifest": "http://hl7.org/fhir/StructureDefinition/DocumentManifest",
  "DocumentReference": "http://hl7.org/fhir/StructureDefinition/DocumentReference",
  "DomainResource": "http://hl7.org/fhir/StructureDefinition/DomainResource",
  "EffectEvidenceSynthesis": "http://hl7.org/fhir/StructureDefinition/EffectEvidenceSynthesis",
  "Encounter": "http://hl7.org/fhir/StructureDefinition/Encounter",
  "Endpoint": "http://hl7.org/fhir/StructureDefinition/Endpoint",
  "EnrollmentRequest": "http://hl7.org/fhir/StructureDefinition/EnrollmentRequest",
  "EnrollmentResponse": "http://hl7.org/fhir/StructureDefinition/EnrollmentResponse",
  "EpisodeOfCare": "http://hl7.org/fhir/StructureDefinition/EpisodeOfCare",
  "EventDefinition": "http://hl7.org/fhir/StructureDefinition/EventDefinition",
  "Evidence": "http://hl7.org/fhir/StructureDefinition/Evidence",
  "EvidenceVariable": "http://hl7.org/fhir/StructureDefinition/EvidenceVariable",
  "ExampleScenario": "http://hl7.org/fhir/StructureDefinition/ExampleScenario",
  "ExplanationOfBenefit": "http://hl7.org/fhir/StructureDefinition/ExplanationOfBenefit",
  "FamilyMemberHistory": "http://hl7.org/fhir/StructureDefinition/FamilyMemberHistory",
  "Flag": "http://hl7.org/fhir/StructureDefinition/Flag",
  "Goal": "http://hl7.org/fhir/StructureDefinition/Goal",
  "GraphDefinition": "http://hl7.org/fhir/StructureDefinition/GraphDefinition",
  "Group": "http://hl7.org/fhir/StructureDefinition/Group",
  "GuidanceResponse": "http://hl7.org/fhir/StructureDefinition/GuidanceResponse",
  "HealthcareService": "http://hl7.org/fhir/StructureDefinition/HealthcareService",
  "ImagingStudy": "http://hl7.org/fhir/StructureDefinition/ImagingStudy",
  "Immunization": "http://hl7.org/fhir/StructureDefinition/Immunization",
  "ImmunizationEvaluation": "http://hl7.org/fhir/StructureDefinition/ImmunizationEvaluation",
  "ImmunizationRecommendation": "http://hl7.org/fhir/StructureDefinition/ImmunizationRecommendation",
  "ImplementationGuide": "http://hl7.org/fhir/StructureDefinition/ImplementationGuide",
  "InsurancePlan": "http://hl7.org/fhir/StructureDefinition/InsurancePlan",
  "Invoice": "http://hl7.org/fhir/StructureDefinition/Invoice",
  "Library": "http://hl7.org/fhir/StructureDefinition/Library",
  "Linkage": "http://hl7.org/fhir/StructureDefinition/Linkage",
  "List": "http://hl7.org/fhir/StructureDefinition/List",
  "Location": "http://hl7.org/fhir/StructureDefinition/Location",
  "Measure": "http://hl7.org/fhir/StructureDefinition/Measure",
  "MeasureReport": "http://hl7.org/fhir/StructureDefinition/MeasureReport",
  "Media": "http://hl7.org/fhir/StructureDefinition/Media",
  "Medication": "http://hl7.org/fhir/StructureDefinition/Medication",
  "MedicationAdministration": "http://hl7.org/fhir/StructureDefinition/MedicationAdministration",
  "MedicationDispense": "http://hl7.org/fhir/StructureDefinition/MedicationDispense",
  "MedicationKnowledge": "http://hl7.org/fhir/StructureDefinition/MedicationKnowledge",
  "MedicationRequest": "http://hl7.org/fhir/StructureDefinition/MedicationRequest",
  "MedicationStatement": "http://hl7.org/fhir/StructureDefinition/MedicationStatement",
  "MedicinalProduct": "http://hl7.org/fhir/StructureDefinition/MedicinalProduct",
  "MedicinalProductAuthorization": "http://hl7.org/fhir/StructureDefinition/MedicinalProductAuthorization",
  "MedicinalProductContraindication": "http://hl7.org/fhir/StructureDefinition/MedicinalProductContraindication",
  "MedicinalProductIndication": "http://hl7.org/fhir/StructureDefinition/MedicinalProductIndication",
  "MedicinalProductIngredient": "http://hl7.org/fhir/StructureDefinition/MedicinalProductIngredient",
  "MedicinalProductInteraction": "http://hl7.org/fhir/StructureDefinition/MedicinalProductInteraction",
  "MedicinalProductManufactured": "http://hl7.org/fhir/StructureDefinition/MedicinalProductManufactured",
  "MedicinalProductPackaged": "http://hl7.org/fhir/StructureDefinition/MedicinalProductPackaged",
  "MedicinalProductPharmaceutical": "http://hl7.org/fhir/StructureDefinition/MedicinalProductPharmaceutical",
  "MedicinalProductUndesirableEffect": "http://hl7.org/fhir/StructureDefinition/MedicinalProductUndesirableEffect",
  "MessageDefinition": "http://hl7.org/fhir/StructureDefinition/MessageDefinition",
  "MessageHeader": "http://hl7.org/fhir/StructureDefinition/MessageHeader",
  "MolecularSequence": "http://hl7.org/fhir/StructureDefinition/MolecularSequence",
  "NamingSystem": "http://hl7.org/fhir/StructureDefinition/NamingSystem",
  "NutritionOrder": "http://hl7.org/fhir/StructureDefinition/NutritionOrder",
  "Observation": "http://hl7.org/fhir/StructureDefinition/Observation",
  "ObservationDefinition": "http://hl7.org/fhir/StructureDefinition/ObservationDefinition",
  "OperationDefinition": "http://hl7.org/fhir/StructureDefinition/OperationDefinition",
  "OperationOutcome": "http://hl7.org/fhir/StructureDefinition/OperationOutcome",
  "Organization": "http://hl7.org/fhir/StructureDefinition/Organization",
  "OrganizationAffiliation": "http://hl7.org/fhir/StructureDefinition/OrganizationAffiliation",
  "Parameters": "http://hl7.org/fhir/StructureDefinition/Parameters",
  "Patient": "http://hl7.org/fhir/StructureDefinition/Patient",
  "PaymentNotice": "http://hl7.org/fhir/StructureDefinition/PaymentNotice",
  "PaymentReconciliation": "http://hl7.org/fhir/StructureDefinition/PaymentReconciliation",
  "Person": "http://hl7.org/fhir/StructureDefinition/Person",
  "PlanDefinition": "http://hl7.org/fhir/StructureDefinition/PlanDefinition",
  "Practitioner": "http://hl7.org/fhir/StructureDefinition/Practitioner",
  "PractitionerRole": "http://hl7.org/fhir/StructureDefinition/PractitionerRole",
  "Procedure": "http://hl7.org/fhir/StructureDefinition/Procedure",
  "Provenance": "http://hl7.org/fhir/StructureDefinition/Provenance",
  "Questionnaire": "http://hl7.org/fhir/StructureDefinition/Questionnaire",
  "QuestionnaireResponse": "http://hl7.org/fhir/StructureDefinition/QuestionnaireResponse",
  "RelatedPerson": "http://hl7.org/fhir/StructureDefinition/RelatedPerson",
  "RequestGroup": "http://hl7.org/fhir/StructureDefinition/RequestGroup",
  "ResearchDefinition": "http://hl7.org/fhir/StructureDefinition/ResearchDefinition",
  "ResearchElementDefinition": "http://hl7.org/fhir/StructureDefinition/ResearchElementDefinition",
  "ResearchStudy": "http://hl7.org/fhir/StructureDefinition/ResearchStudy",
  "ResearchSubject": "http://hl7.org/fhir/StructureDefinition/ResearchSubject",
  "RiskAssessment": "http://hl7.org/fhir/StructureDefinition/RiskAssessment",
  "RiskEvidenceSynthesis": "http://hl7.org/fhir/StructureDefinition/RiskEvidenceSynthesis",
  "Schedule": "http://hl7.org/fhir/StructureDefinition/Schedule",
  "SearchParameter": "http://hl7.org/fhir/StructureDefinition/SearchParameter",
  "ServiceRequest": "http://hl7.org/fhir/StructureDefinition/ServiceRequest",
  "Slot": "http://hl7.org/fhir/StructureDefinition/Slot",
  "Specimen": "http://hl7.org/fhir/StructureDefinition/Specimen",
  "SpecimenDefinition": "http://hl7.org/fhir/StructureDefinition/SpecimenDefinition",
  "StructureDefinition": "http://hl7.org/fhir/StructureDefinition/StructureDefinition",
  "StructureMap": "http://hl7.org/fhir/StructureDefinition/StructureMap",
  "Subscription": "http://hl7.org/fhir/StructureDefinition/Subscription",
  "Substance": "http://hl7.org/fhir/StructureDefinition/Substance",
  "SubstanceNucleicAcid": "http://hl7.org/fhir/StructureDefinition/SubstanceNucleicAcid",
  "SubstancePolymer": "http://hl7.org/fhir/StructureDefinition/SubstancePolymer",
  "SubstanceProtein": "http://hl7.org/fhir/StructureDefinition/SubstanceProtein",
  "SubstanceReferenceInformation": "http://hl7.org/fhir/StructureDefinition/SubstanceReferenceInformation",
  "SubstanceSourceMaterial": "http://hl7.org/fhir/StructureDefinition/SubstanceSourceMaterial",
  "SubstanceSpecification": "http://hl7.org/fhir/StructureDefinition/SubstanceSpecification",
  "SupplyDelivery": "http://hl7.org/fhir/StructureDefinition/SupplyDelivery",
  "SupplyRequest": "http://hl7.org/fhir/StructureDefinition/SupplyRequest",
  "Task": "http://hl7.org/fhir/StructureDefinition/Task",
  "TerminologyCapabilities": "http://hl7.org/fhir/StructureDefinition/TerminologyCapabilities",
  "TestReport": "http://hl7.org/fhir/StructureDefinition/TestReport",
  "TestScript": "http://hl7.org/fhir/StructureDefinition/TestScript",
  "ValueSet": "http://hl7.org/fhir/StructureDefinition/ValueSet",
  "VerificationResult": "http://hl7.org/fhir/StructureDefinition/VerificationResult",
  "VisionPrescription": "http://hl7.org/fhir/StructureDefinition/VisionPrescription",
  "Element": "http://hl7.org/fhir/StructureDefinition/Element",
  "BackboneElement": "http://hl7.org/fhir/StructureDefinition/BackboneElement",
  "base64Binary": "http://hl7.org/fhir/StructureDefinition/base64Binary",
  "boolean": "http://hl7.org/fhir/StructureDefinition/boolean",
  "canonical": "http://hl7.org/fhir/StructureDefinition/canonical",
  "code": "http://hl7.org/fhir/StructureDefinition/code",
  "date": "http://hl7.org/fhir/StructureDefinition/date",
  "dateTime": "http://hl7.org/fhir/StructureDefinition/dateTime",
  "decimal": "http://hl7.org/fhir/StructureDefinition/decimal",
  "id": "http://hl7.org/fhir/StructureDefinition/id",
  "instant": "http://hl7.org/fhir/StructureDefinition/instant",
  "integer": "http://hl7.org/fhir/StructureDefinition/integer",
  "markdown": "http://hl7.org/fhir/StructureDefinition/markdown",
  "oid": "http://hl7.org/fhir/StructureDefinition/oid",
  "positiveInt": "http://hl7.org/fhir/StructureDefinition/positiveInt",
  "string": "http://hl7.org/fhir/StructureDefinition/string",
  "time": "http://hl7.org/fhir/StructureDefinition/time",
  "unsignedInt": "http://hl7.org/fhir/StructureDefinition/unsignedInt",
  "uri": "http://hl7.org/fhir/StructureDefinition/uri",
  "url": "http://hl7.org/fhir/StructureDefinition/url",
  "uuid": "http://hl7.org/fhir/StructureDefinition/uuid",
  "xhtml": "http://hl7.org/fhir/StructureDefinition/xhtml",
  "Address": "http://hl7.org/fhir/StructureDefinition/Address",
  "Age": "http://hl7.org/fhir/StructureDefinition/Age",
  "Annotation": "http://hl7.org/fhir/StructureDefinition/Annotation",
  "Attachment": "http://hl7.org/fhir/StructureDefinition/Attachment",
  "CodeableConcept": "http://hl7.org/fhir/StructureDefinition/CodeableConcept",
  "Coding": "http://hl7.org/fhir/StructureDefinition/Coding",
  "ContactDetail": "http://hl7.org/fhir/StructureDefinition/ContactDetail",
  "ContactPoint": "http://hl7.org/fhir/StructureDefinition/ContactPoint",
  "Contributor": "http://hl7.org/fhir/StructureDefinition/Contributor",
  "Count": "http://hl7.org/fhir/StructureDefinition/Count",
  "DataRequirement": "http://hl7.org/fhir/StructureDefinition/DataRequirement",
  "Distance": "http://hl7.org/fhir/StructureDefinition/Distance",
  "Dosage": "http://hl7.org/fhir/StructureDefinition/Dosage",
  "Duration": "http://hl7.org/fhir/StructureDefinition/Duration",
  "ElementDefinition": "http://hl7.org/fhir/StructureDefinition/ElementDefinition",
  "Expression": "http://hl7.org/fhir/StructureDefinition/Expression",
  "Extension": "http://hl7.org/fhir/StructureDefinition/Extension",
  "HumanName": "http://hl7.org/fhir/StructureDefinition/HumanName",
  "Identifier": "http://hl7.org/fhir/StructureDefinition/Identifier",
  "MarketingStatus": "http://hl7.org/fhir/StructureDefinition/MarketingStatus",
  "Meta": "http://hl7.org/fhir/StructureDefinition/Meta",
  "Money": "http://hl7.org/fhir/StructureDefinition/Money",
  "Narrative": "http://hl7.org/fhir/StructureDefinition/Narrative",
  "ParameterDefinition": "http://hl7.org/fhir/StructureDefinition/ParameterDefinition",
  "Period": "http://hl7.org/fhir/StructureDefinition/Period",
  "Population": "http://hl7.org/fhir/StructureDefinition/Population",
  "ProdCharacteristic": "http://hl7.org/fhir/StructureDefinition/ProdCharacteristic",
  "ProductShelfLife": "http://hl7.org/fhir/StructureDefinition/ProductShelfLife",
  "Quantity": "http://hl7.org/fhir/StructureDefinition/Quantity",
  "Range": "http://hl7.org/fhir/StructureDefinition/Range",
  "Ratio": "http://hl7.org/fhir/StructureDefinition/Ratio",
  "Reference": "http://hl7.org/fhir/StructureDefinition/Reference",
  "RelatedArtifact": "http://hl7.org/fhir/StructureDefinition/RelatedArtifact",
  "SampledData": "http://hl7.org/fhir/StructureDefinition/SampledData",
  "Signature": "http://hl7.org/fhir/StructureDefinition/Signature",
  "SubstanceAmount": "http://hl7.org/fhir/StructureDefinition/SubstanceAmount",
  "Timing": "http://hl7.org/fhir/StructureDefinition/Timing",
  "TriggerDefinition": "http://hl7.org/fhir/StructureDefinition/TriggerDefinition",
  "UsageContext": "http://hl7.org/fhir/StructureDefinition/UsageContext",
  "AccessPolicyV2": "https://iguhealth.com/StructureDefinition/AccessPolicyV2",
  "AccessPolicy": "https://iguhealth.com/StructureDefinition/AccessPolicy",
  "ClientApplication": "https://iguhealth.com/StructureDefinition/ClientApplication",
  "IdentityProvider": "https://iguhealth.com/StructureDefinition/IdentityProvider",
  "Membership": "https://iguhealth.com/StructureDefinition/Membership",
  "MessageBroker": "https://iguhealth.com/StructureDefinition/MessageBroker",
  "MessageTopic": "https://iguhealth.com/StructureDefinition/MessageTopic"
} as Record<string, string>