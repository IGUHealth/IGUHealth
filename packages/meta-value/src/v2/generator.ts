import { traversalBottomUp } from "@iguhealth/codegen";
import { ElementDefinition, uri } from "@iguhealth/fhir-types/r4/types";
import { FHIR_VERSION, Resource } from "@iguhealth/fhir-types/versions";

interface TypeChoiceNode {
  _type_: "typechoice";
  types: Record<uri, number>;
}

interface SingularNode {
  _type_: "meta";
  type: uri;
  cardinality: "array" | "single";
  properties?: Record<string, number>;
}

type MetaNode = SingularNode | TypeChoiceNode;

function determineIsTypeChoice(element: ElementDefinition): boolean {
  return (element.type ?? []).length > 1;
}

interface MetaV2Compiled {
  [key: string]: Array<MetaNode>;
}

type ElementPath = string & { _elementPathBrand: unknown };
type ElementPathType = string & { _elementPathTypeBrand: unknown };

function combineWithPath(path: ElementPath, type: uri): ElementPathType {
  return `${path}:${type}` as ElementPathType;
}

// Generate indices around the element path and it's type which correlate to location in the meta.
function preGenerateIndices(
  sd: Resource<FHIR_VERSION, "StructureDefinition">,
): Record<ElementPath | ElementPathType, number> {
  let curIndex = 0;
  // Record
  const indices: Record<ElementPath | ElementPathType, number> = {};

  for (let i = 0; i < (sd.snapshot?.element ?? []).length; i++) {
    const element = sd.snapshot?.element[i];
    if (!element) throw new Error("No Element");
    const types = element?.type ?? [];
    indices[element.path as ElementPath] = curIndex;
    if (determineIsTypeChoice(element)) {
      for (const type of types) {
        indices[combineWithPath(element.path as ElementPath, type.code)] =
          curIndex;
      }
    }

    curIndex++;
  }

  return indices;
}

// function to capaitalize a string
function capitalize(str: string) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}

function getElementField(element: ElementDefinition, type?: string) {
  const path = element.path;
  const pathSplit = path.split(".");
  let field = pathSplit[pathSplit.length - 1];
  if (type) {
    field = field.replace("[x]", capitalize(type));
  }
  return field;
}

function SDToMetaData(sd: Resource<FHIR_VERSION, "StructureDefinition">) {
  const indices = preGenerateIndices(sd);
  console.log(indices);
  const metaInfo: Array<MetaNode> = [
    ...new Array(Math.max(...Object.values(indices))),
  ];

  traversalBottomUp<{ index: number; field: string }>(
    sd,
    (element: ElementDefinition, children) => {
      if (element.contentReference) {
        const index =
          indices[element.contentReference as unknown as ElementPath];
        return [
          {
            index,
            field: getElementField(element),
          },
        ];
      }

      if (determineIsTypeChoice(element)) {
        for (const type of element.type ?? []) {
          const indexKey = combineWithPath(
            element.path as ElementPath,
            type.code,
          );

          const index = indices[indexKey];
          metaInfo[index] = {
            _type_: "meta",
            type: type.code as uri,
            cardinality: element.max === "1" ? "single" : "array",
            properties:
              children.length === 0
                ? undefined
                : children.reduce((acc: Record<string, number>, k) => {
                    acc[k.field] = k.index;
                    return acc;
                  }, {}),
          };
        }

        // For Type choices include a typechoice node which points to the types loc.

        metaInfo[indices[element.path as ElementPath]] = {
          _type_: "typechoice",
          types:
            element.type?.reduce((acc: Record<uri, number>, type) => {
              acc[type.code] =
                indices[
                  combineWithPath(element.path as ElementPath, type.code)
                ];
              return acc;
            }, {}) ?? {},
        };
      } else {
        const index = indices[element.path as ElementPath];
        metaInfo[index] = {
          _type_: "meta",
          type: element.type?.[0].code as uri,
          cardinality: element.max === "1" ? "single" : "array",
          properties:
            children.length === 0
              ? undefined
              : children.reduce((acc: Record<string, number>, k) => {
                  acc[k.field] = k.index;
                  return acc;
                }, {}),
        };
      }

      return (element.type ?? []).map((type) => {
        return {
          index: indices[element.path as ElementPath],
          field: getElementField(element, type.code),
        };
      });
    },
  );

  return metaInfo;
}

export function generateMetaData<Version extends FHIR_VERSION>(
  sds: Resource<Version, "StructureDefinition">[],
): MetaV2Compiled {
  const meta: MetaV2Compiled = {};

  for (const sd of sds) {
    meta[sd.type] = SDToMetaData(sd);
  }

  return meta;
}
