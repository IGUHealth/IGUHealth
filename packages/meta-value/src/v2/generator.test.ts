import { expect, test } from "@jest/globals";
import path from "path";
import { fileURLToPath } from "url";

import { loadArtifacts } from "@iguhealth/artifacts";
import { StructureDefinition } from "@iguhealth/fhir-types/r4/types";
import { R4 } from "@iguhealth/fhir-types/versions";

import { generateMetaData } from "./generator.js";

const sds: StructureDefinition[] = loadArtifacts({
  fhirVersion: R4,
  loadDevelopmentPackages: true,
  resourceType: "StructureDefinition",
  packageLocation: path.join(fileURLToPath(import.meta.url), "../../"),
});

test("Generate R4 Patient", async () => {
  expect(
    generateMetaData(sds.filter((sd) => sd.type === "Patient")),
  ).toMatchSnapshot();
});

test("Generate R4 Questionnaire", async () => {
  expect(
    generateMetaData(sds.filter((sd) => sd.type === "Questionnaire")),
  ).toMatchSnapshot();
});
