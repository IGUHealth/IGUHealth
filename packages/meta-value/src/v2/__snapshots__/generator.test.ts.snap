// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate R4 Patient 1`] = `
{
  "Patient": [
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": {
        "active": 10,
        "address": 16,
        "birthDate": 14,
        "communication": 31,
        "contact": 20,
        "contained": 6,
        "deceasedBoolean": 15,
        "deceasedDateTime": 15,
        "extension": 7,
        "gender": 13,
        "generalPractitioner": 37,
        "id": 1,
        "identifier": 9,
        "implicitRules": 3,
        "language": 4,
        "link": 39,
        "managingOrganization": 38,
        "maritalStatus": 17,
        "meta": 2,
        "modifierExtension": 8,
        "multipleBirthBoolean": 18,
        "multipleBirthInteger": 18,
        "name": 11,
        "photo": 19,
        "telecom": 12,
        "text": 5,
      },
      "type": undefined,
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "http://hl7.org/fhirpath/System.String",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "Meta",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "uri",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "code",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "Narrative",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Resource",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Identifier",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "boolean",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "HumanName",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "ContactPoint",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "code",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "date",
    },
    {
      "_type_": "typechoice",
      "types": {
        "boolean": 15,
        "dateTime": 15,
      },
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Address",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "CodeableConcept",
    },
    {
      "_type_": "typechoice",
      "types": {
        "boolean": 18,
        "integer": 18,
      },
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Attachment",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": {
        "address": 27,
        "extension": 22,
        "gender": 28,
        "id": 21,
        "modifierExtension": 23,
        "name": 25,
        "organization": 29,
        "period": 30,
        "relationship": 24,
        "telecom": 26,
      },
      "type": "BackboneElement",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "http://hl7.org/fhirpath/System.String",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "CodeableConcept",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "HumanName",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "ContactPoint",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "Address",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "code",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "Reference",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "Period",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": {
        "extension": 33,
        "id": 32,
        "language": 35,
        "modifierExtension": 34,
        "preferred": 36,
      },
      "type": "BackboneElement",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "http://hl7.org/fhirpath/System.String",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "CodeableConcept",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "boolean",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Reference",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "Reference",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": {
        "extension": 41,
        "id": 40,
        "modifierExtension": 42,
        "other": 43,
        "type": 44,
      },
      "type": "BackboneElement",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "http://hl7.org/fhirpath/System.String",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "array",
      "properties": undefined,
      "type": "Extension",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "Reference",
    },
    {
      "_type_": "meta",
      "cardinality": "single",
      "properties": undefined,
      "type": "code",
    },
  ],
}
`;
